src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -60659,5 +60675,1033 @@ public void TestMeetsAgree()
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void NullableT_AllMembers()
+        {
+            var source =
+@"class C<T> where T : struct
+{
+    static void F1(T? t1)
+    {
+        _ = t1.HasValue;
+    }
+    static void F2(T? t2)
+    {
+        _ = t2.Value; // 1
+    }
+    static void F3(T? t3)
+    {
+        _ = t3.GetValueOrDefault();
+    }
+    static void F4(T? t4)
+    {
+        _ = t4.GetHashCode();
+    }
+    static void F5(T? t5)
+    {
+        _ = t5.ToString();
+    }
+    static void F6(T? t6)
+    {
+        _ = t6.Equals(t6);
+    }
+}";
+            var comp = CreateCompilation(source, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                // (9,13): warning CS8629: Nullable value type may be null.
+                //         _ = t2.Value; // 1
+                Diagnostic(ErrorCode.WRN_NullableValueTypeMayBeNull, "t2").WithLocation(9, 13));
+        }
+
+        [Fact]
+        public void NullableT_Using()
+        {
+            var source =
+@"using System;
+struct S : IDisposable
+{
+    void IDisposable.Dispose() { }
+}
+class Program
+{
+    static void F1(S? s)
+    {
+        using (s) { }
+        _ = s.Value; // 1
+    }
+    static void F2<T>(T? t) where T : struct, IDisposable
+    {
+        using (t) { }
+        _ = t.Value; // 2
+    }
+}";
+            var comp = CreateCompilation(source, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                // (11,13): warning CS8629: Nullable value type may be null.
+                //         _ = s.Value; // 1
+                Diagnostic(ErrorCode.WRN_NullableValueTypeMayBeNull, "s").WithLocation(11, 13),
+                // (16,13): warning CS8629: Nullable value type may be null.
+                //         _ = t.Value; // 2
+                Diagnostic(ErrorCode.WRN_NullableValueTypeMayBeNull, "t").WithLocation(16, 13));
+        }
+
+        [WorkItem(31503, "https://github.com/dotnet/roslyn/issues/31503")]
+        [Fact]
+        public void NullableT_ForEach()
+        {
+            var source =
+@"using System.Collections;
+using System.Collections.Generic;
+struct S : IEnumerable
+{
+    public IEnumerator GetEnumerator() => throw null;
+}
+class Program
+{
+    static void F1(S? s)
+    {
+        foreach (var i in s) // 1
+            ;
+        foreach (var i in s)
+            ;
+    }
+    static void F2<T, U>(T? t) where T : struct, IEnumerable<U>
+    {
+        foreach (var i in t) // 2
+            ;
+        foreach (var i in t)
+            ;
+    }
+}";
+            var comp = CreateCompilation(source, options: WithNonNullTypesTrue());
+            // https://github.com/dotnet/roslyn/issues/31503: Report warning for `.Value.GetEnumerator()` calls.
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void NullableT_IndexAndRange()
```

Comment:
Are we expecting to get warnings here, or is this the actual final behavior? If the former, consider logging a bug.

Reply:
There shouldn't be any warnings here. (There were warnings in an earlier iteration of this change.)

---
In reply to: [238833118](https://github.com/dotnet/roslyn/pull/31499#discussion_r238833118) [](ancestors = 238833118)